dask/array/core.py

```diff
@@ -2833,20 +2833,21 @@ def where(condition, x=None, y=None):
     if x is None or y is None:
         raise TypeError(where_error_message)
 
-    x = asarray(x)
-    y = asarray(y)
-
-    shape = broadcast_shapes(x.shape, y.shape)
-    dtype = np.promote_types(x.dtype, y.dtype)
+    if np.isscalar(condition):
+        # Match the dtype inference logic from elemwise
+        vals = [np.empty((1,) * a.ndim, dtype=a.dtype)
+                if not is_scalar_for_elemwise(a) else a
+                for a in (x, y)]
+        dtype = apply_infer_dtype(np.where, [condition] + vals, {}, 'where', suggest_dtype=False)
```

Comment:
No, that doesn't work either (apart from it needing a variadic list, not a generator expression), because numpy scalar have a `dtype` attribute:
```python
>>> x = np.float64(1.0)
>>> y = np.array([1, 2], np.float32)
>>> np.result_type(*(np.empty((1,), dtype=a.dtype) if hasattr(a, 'dtype') else a
                     for a in (x, y)))
dtype('float64')
>>> np.where(True, x, y).dtype
dtype('float32')
```
I'm trying to ensure that the scalar-conditional optimisation should always give the same dtype as the generic path through `elemwise`. When I get a chance (probably only tomorrow) I'll try pulling out the `elemwise` dtype logic into a function I can just call here.